internal/graph/checkingresourcestream.go

```diff
@@ -0,0 +1,537 @@
+package graph
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/authzed/spicedb/internal/dispatch"
+	"github.com/authzed/spicedb/internal/graph/computed"
+	"github.com/authzed/spicedb/pkg/datastore"
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+	"github.com/authzed/spicedb/pkg/util"
+)
+
+// possibleResource is a resource that was returned by reachable resources and, after processing,
+// may be returned by the lookup resources stream.
+type possibleResource struct {
+	// reachableResult is the result for this resource from the reachable resources stream.
+	reachableResult *v1.DispatchReachableResourcesResponse
+
+	// lookupResult is the result to be published by LookupResources, if the resource is actually
+	// permissioned. Will be nil before processing and nil after processing IF the resource needed
+	// to be checked and the check failed.
+	lookupResult *v1.DispatchLookupResourcesResponse
+
+	// orderingIndex is the index of the resource result as returned by reachable resources. Used to
+	// maintain strict publishing order of results.
+	orderingIndex uint64
+}
+
+// resourceQueue is a queue for managing of possibleResources through the various states of the stream (queueing, processing and publishing).
+type resourceQueue struct {
+	ctx  context.Context
+	lock sync.Mutex
+
+	// toProcess are those resources (keyed by orderingIndex) that have not yet been processed (checked).
+	toProcess map[uint64]possibleResource
+
+	// toPublish are those resources (keyed by orderingIndex) that have been processed and
+	// are ready for publishing. Note that resources whose Check calls showed NO_PERMISSION
+	// will *also* be in this map, just with lookupResult set to nil. This is done to ensure
+	// strict ordering of publishing.
+	toPublish map[uint64]possibleResource
+
+	// beingProcessed are those resources (keyed by orderingIndex) that are currently being processed.
+	beingProcessed map[uint64]possibleResource
+}
+
+// addPossibleResources queues resources for processing.
+func (rq *resourceQueue) addPossibleResources(prs ...possibleResource) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+
+	for _, pr := range prs {
+		rq.toProcess[pr.orderingIndex] = pr
+	}
+}
+
+// updateToBePublished marks resources as ready for publishing.
+func (rq *resourceQueue) updateToBePublished(prs ...possibleResource) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+
+	for _, pr := range prs {
+		rq.toPublish[pr.orderingIndex] = pr
+		delete(rq.beingProcessed, pr.orderingIndex)
+	}
+}
+
+// markResourceCompleted marks that all work has been completed on the given resources.
+func (rq *resourceQueue) markResourceCompleted(prs ...possibleResource) {
```

Comment:
I don't see any call to this method with more than 1 resource. I'm not sure if the compiler is clever enough to avoid the allocation, but if there is no use we should probably remove it anyway, and turn it into a variadic when the need arises.